### PR TITLE
Bugfix: Instruction name / pulse library name mismatch

### LINKED ISSUE #4047: Add id field for unique identification of Instructions and PulseLibraryItem in pulse
### What is the expected enhancement?

As suggested by @lcapelluto : Add some form of generated unique identifier `id` field for `SamplePulse`/`Instruction`/`PulseLibraryItem`/`...` in pulse, so that they can be identified using this instead of the `name` field.

### Purpose

As far as I am aware, the only way to identify a `SamplePulse`/`Instruction`/`PulseLibraryItem`/`...` in pulse is through the `name` field. This has caused a recent issue in `Aer` as the names of `PulseLibraryItem`s now get a hash added at the end, and as such there ends up being a name mismatch between an `Instruction` in an experiment to play a sample pulse, and the name of the sample pulse itself in an assembled `PulseQobj`'s `pulse_library` (please correct me if the terminology i'm using is incorrect). E.g. if a sample pulse has name `rabi_pulse`, the `Instruction` to play the pulse will inherit the name `rabi_pulse`, but the name of the pulse in the `pulse_library` will be something like `rabi_pulse--83975728198`. As such, relating the `Instruction` to the sample pulse in the `pulse_library` using the `name` requires ignoring the `--83975728198`, which to work requires assumptions on what a user will "typically" enter into the `name` field.

### PR DESCRIPTION
**Pull request by lcapelluto**
### Summary
Fix a critical bug where the Qobj sample pulse instruction did not have the same name as the pulse library pulse

Fix #4047

### CODE REVIEW

File: qiskit/assembler/assemble_schedules.py [hunk 1]
```diff
@@ -171,40 +172,26 @@ def _assemble_instructions(
...
-                    command=SamplePulse(name=name, samples=instruction.command.samples),
-                    name=instruction.name,
-                    channel=instruction.channels[0])
-            # add samples to pulse library
-            user_pulselib[name] = instruction.command
+            name = hashlib.sha256(instruction.pulse.samples).hexdigest()
+            instruction =  Play(SamplePulse(name=name, samples=instruction.pulse.samples),
+                                channel=instruction.channel,
+                                name=name)
+            user_pulselib[name] = instruction.pulse.samples
```
Comment (lcapelluto) [inline qiskit/assembler/assemble_schedules.py]: If we're worried about hash collisions, could have something like:
```
            if name in user_pulselib and not np.allclose(user_pulselib[name], instruction.pulse.samples):
                name += str(instruction.pulse.id)
```
but `pulse.id` will be implemented in another PR

File: qiskit/assembler/assemble_schedules.py [hunk 2]
```diff
@@ -171,40 +172,26 @@ def _assemble_instructions(
...
-            if name in user_pulselib and instruction.command != user_pulselib[name]:
-                name = "{0}-{1:x}".format(name, hash(instruction.command.samples.tostring()))
-                instruction = PulseInstruction(
-                    command=SamplePulse(name=name, samples=instruction.command.samples),
-                    name=instruction.name,
-                    channel=instruction.channels[0])
-            # add samples to pulse library
-            user_pulselib[name] = instruction.command
+            name = hashlib.sha256(instruction.pulse.samples).hexdigest()
+            instruction =  Play(SamplePulse(name=name, samples=instruction.pulse.samples),
```
Comment (DanPuzzuoli) [inline qiskit/assembler/assemble_schedules.py]:
```suggestion
            instruction = Play(SamplePulse(name=name, samples=instruction.pulse.samples),
```

Comment (DanPuzzuoli) [reply qiskit/assembler/assemble_schedules.py]: I think this will fix the lint issue

File: qiskit/assembler/assemble_schedules.py (hunk 1 above)
Comment (DanPuzzuoli) [reply qiskit/assembler/assemble_schedules.py]: sounds good!

File: qiskit/assembler/assemble_schedules.py (hunk 2 above)
Comment (DanPuzzuoli) [reply qiskit/assembler/assemble_schedules.py]: Actually my bad: just realized removing that space will also require realigning the subsequent lines as well.

Comment (lcapelluto) [reply qiskit/assembler/assemble_schedules.py]: oh thank you:)
